Marlin/src/module/stepper.cpp

```diff
@@ -225,8 +231,8 @@ xyze_long_t Stepper::delta_error{0};
 xyze_long_t Stepper::advance_dividend{0};
 uint32_t Stepper::advance_divisor = 0,
          Stepper::step_events_completed = 0, // The number of step events executed in the current block
-         Stepper::accelerate_until,          // The count at which to stop accelerating
-         Stepper::decelerate_after,          // The count at which to start decelerating
+         Stepper::accelerate_before,         // The count at which to start cruising
```

Comment:
Stepper.cpp uses `step_events_completed` and that's incremented _before_ it is being checked against `accelerate_until`/`accelerate_before`. By the time we are checking the relevant step is complete, which is probably why there was the off-by-one in the first place. Because of that I don't clearly see new names as better, just different (it will add some extra merge conflicts with a future change I'm working on but that's a minor thing). Does the new name help you understand the code better?

Reply:
The naming of these variables should be consistent with their meaning, and if the step index being compared against these values is not the current but the next step, that should be reflected in the naming. However, that does not seem to be the case.

Although `step_events_completed` is being used for comparison after a previous `pulse_phase_isr`, the variable  is still and always a 0-based index, only ending up at `step_event_count` (past the end) after the last step has been executed in the pulse phase.

So in the comparison at line 2309 `step_events_completed < accelerate_before` the suffix "before" makes sense. I also considered calling it "`cruise_start`" since that's how the initial deceleration index is named. Would this rename make it a little clearer?

Reply:
We're comparing whether `step_events_completed < accelerate_before` is clearer than `step_events_completed < accelerate_until`. You have a point that in `step_events_completed < accelerate_before` the suffix "before" makes sense, just looking at the comparison. However, the code that follows does the preparation for what the next step is and that step runs on the next ISR.

Say we have a segment with 5 acceleration steps to do. The `accelerate_before/until` will have a value of 5. The first step is prepared by the previous segment's `block_phase_isr` to run at `initial_rate` and have '1/2' its interval into `acceleration_time` (therefore we can kinda call it an acceleration step). Now after a `pulse_phase_isr` `step_events_completed` is 1. We check and `step_events_completed < accelerate_before/until` so we prepare another acceleration step. `pulse_phase_isr` runs and now `step_events_completed` is 2. We check and `step_events_completed < accelerate_before/until` so we prepare another acceleration step. ISR runs and now `step_events_completed` is 2...
..
ISR runs and now `step_events_completed` is 5. We check and `step_events_completed < accelerate_before/until` is false.

> during block_phase_isr the step_events_completed value should be considered to be the zero-based index of the step currently being examined

That's not how my brain reads `step_events_completed` :). It comes naturally to me to increment a variable and hard to try to switch its zero/one index. And I personally read `accelerate_until` as keep accelerating until we've completed this many steps (which works in the context of checking it against `step_events_completed` when setting up what the next step will do). 

That said, as long as the code is correct, I don't really mind `accelerate_before`/`until` much, other than `block->accelerate_before = accelerate_steps;` reads wrong to me.

> I also considered calling it "cruise_start" since that's how the initial deceleration index is named.

Not good since we often go from accelerate straight into decelerate.

Reply:
The "next step index" always corresponds to "last run step index plus one" which corresponds to "number of steps completed." These are all equivalent to one another, but as you point out we only have one ugly variable name. At the time when `step_events_completed` equals 2, there have been 2 steps completed, indexes 0 and 1, and so the next step index is 2. So outside of `pulse_phase_isr` we expect the "count" to correspond to the index of the next step. And when it equals or exceeds the total number of steps it acts as a special indicator that we're "done" with the block.